mlflow/sklearn/__init__.py

```diff
@@ -994,3 +986,79 @@ def patched_fit(original, self, *args, **kwargs):
                 safe_patch(
                     FLAVOR_NAME, class_def, func_name, patched_fit, manage_run=True,
                 )
+
+
+def eval_and_log_metrics(*, model, X, y_true, prefix, sample_weight=None):
+    """
+    Computes and logs metrics (and artifacts) for the given model and labeled dataset.
+    The metrics/artifacts mirror what is auto-logged when training a model
+    (see mlflow.sklearn.autolog).
+
+    ** Example **
+
+    .. code-block:: python
+
+        from sklearn.linear_model import LinearRegression
+        import mlflow
+
+        # enable autologging
+        mlflow.sklearn.autolog()
+
+        # prepare training data
+        X = np.array([[1, 1], [1, 2], [2, 2], [2, 3]])
+        y = np.dot(X, np.array([1, 2])) + 3
+
+        # prepare evaluation data
+        X_eval = np.array([[3, 3], [3, 4]])
+        y_eval = np.dot(X_eval, np.array([1,2])) + 3
+
+        # train a model
+        model = LinearRegression()
+        with mlflow.start_run() as run:
+            model.fit(X, y)
+            (metrics, artifacts) = mlflow.sklearn.eval_and_log_metrics(model=model, X=X_eval,
+                                                                       y_true=y_eval, prefix="val_")
+
+
+    Each metric's and artifact's name is prefixed with `prefix`, e.g., in the previous example the
+    metrics and artifacts are named 'val_XXXXX'. Note that training-time metrics are auto-logged
+    as 'training_XXXXX'. Metrics and artifacs are logged under the current run if one exists,
+    otherwise a new run is started and left open.
+
+    Raises an error if:
+    - prefix is empty
+    - model is not an sklearn estimator or does not support the 'predict' method
+
+    :param model: The model to be evaluated.
+    :param X: The features for the evaluation dataset.
+    :param y_true: The labels for the evaluation dataset.
+    :param prefix: Prefix used to name metrics and artifacts.
+    :param sample_weight: Per-sample weights to apply in the computation of metrics/artifacts.
+    :return: A tuple (metrics, artifacts), where metrics is a dict of the logged metrics, and
+             artifacts a list of the logged artifact paths.
+    """
+    from mlflow.sklearn.utils import _log_estimator_content
+    from sklearn.base import BaseEstimator
+
+    if prefix is None or prefix == "":
+        raise ValueError("must specify a non-empty prefix")
+
+    if not isinstance(model, BaseEstimator):
+        raise ValueError("not a sklearn model")
+
+    if not hasattr(model, "predict"):
+        raise ValueError("model does not support predictions")
+
+    active_run = mlflow.active_run()
+    run = active_run if active_run is not None else mlflow.start_run()
+
+    (metrics, artifacts) = _log_estimator_content(
+        estimator=model,
+        run_id=run.info.run_id,
+        prefix=prefix,
+        X=X,
+        y_true=y_true,
+        sample_weight=sample_weight,
+    )
+
+    return (metrics, artifacts)
```

Comment:
I still prefer returning only `metrics`. If users need artifact paths, they can lookup the run info. I think that will match what users guess based on the method name. If we return a tuple, most users need to lookup the API doc to understand the return value structure, which is also easy to forget.

Reply:
Is there any reasonable use case for `artifacts` 🤔 ?

Reply:
For logging I think there is, in order to maintain symmetry with training auto-logging. For programmatic access, it's a fair question. If there are no precedents in MLflow's usage patterns then we can remove the artifacts return and simplify following mengxr's suggestion. Wdyt?

Reply:
One use case (which I think is minor) that I could come up with is:

```python
metrics, artifact_names = mlflow.sklearn.eval_and_log_metrics(...)
artifact_paths = ...

# I want to see the results now on the notebook, not on MLflow UI!
import matplotlib.pyplot as plt

for p in artifact_paths:
    plt.imread(p)
    plt.show()
```

Reply:
However, I'd vote for returning only metrics for simplicity.

Reply:
Done.